src/mame/f256/tiny_vicky.cpp

```diff
@@ -0,0 +1,613 @@
+    memcpy(tiles, m_videoram_ptr + tilemapAddress + (tilemapWindowX / tileSize) * 2 + (tileRow + 0) * tilemapWidth * 2, tlmSize);
+
+    // cache of tilesetPointers
+    int tilesetPointers[8];
+    int strides[8];
+    for (int i = 0; i < 8; i++)
+    {
+        tilesetPointers[i] = (m_iopage0_ptr[0xD280 - 0xC000 + i * 4] + (m_iopage0_ptr[0xD280 - 0xC000 + i * 4 + 1] << 8) +
+                             (m_iopage0_ptr[0xD280 - 0xC000 + i * 4 + 2] << 16)) & 0x3F'FFFF;
+        uint8_t tilesetConfig = m_iopage0_ptr[0xD280 - 0xC000 + i * 4 + 3];
+        strides[i] = (tilesetConfig & 8) != 0 ? strideLine : tileSize;
+    }
+    for (int i = 0; i < tilemapItemCount; i++)
+    {
+        uint8_t tile = tiles[i * 2];
+        uint8_t tilesetReg = tiles[i * 2 + 1];
+        uint8_t tileset = tilesetReg & 7;
+
+        // tileset
+        int tilesetPointer = tilesetPointers[tileset];
+        int strideX = strides[tileset];
+        if (strideX == tileSize)
+        {
+            tilesetOffsets[i] = tilesetPointer + (tile % 16) * tileSize * tileSize + (tile / 16) * tileSize * tileSize * 16 + tileYOffset * tileSize;
+        }
+        else
+        {
+            tilesetOffsets[i] = tilesetPointer + ((tile / 16) * strideX * tileSize + (tile % 16) * tileSize) + tileYOffset * strideX;
+        }
+    }
+
+    // alternate display style - avoids repeating the loop so often
+    int startTileX = (borderXSize + tileXOffset) / tileSize;
+    int endTileX = (width/2 - borderXSize + tileXOffset) / tileSize + 1;
+    int startOffset = (borderXSize + tilemapWindowX) % tileSize;
+    int x = borderXSize;
+    //uint8_t tilepix[tileSize];
+    rgb_t clr_val = 0;
+    for (int t = startTileX; t < endTileX; t++)
+    {
+        // The (mode==0 ? 1 : 3) below is to take an FPGA but in the F256Jr into account
+        uint8_t tilesetReg = tiles[t * 2 + 3];
+        uint8_t lut_index = (tilesetReg & 0x38) >> 3;
+        //int lutAddress = MemoryMap.GRP_LUT_BASE_ADDR - VICKY.StartAddress + lutIndex * 1024;
+        //int tilesetOffsetAddress = tilesetOffsets[t];  // + startOffset
+        //memcpy(tilepix, m_videoram_ptr + tilesetOffsets[t], tileSize);
+        do
+        {
+            uint8_t pixVal = m_videoram_ptr[tilesetOffsets[t] + startOffset];  // tilepix[startOffset];
+            if (pixVal > 0)
+            {
+                clr_val = get_lut_value(lut_index, pixVal, enable_gamma);
+                row[x] = clr_val;
+                row[x+1] = clr_val;
+            }
+            startOffset++;
+            //tilesetOffsetAddress++;
+            x+=2;
+        } while (startOffset != tileSize && x < maxX);
+        startOffset = 0;
+        if (x == maxX)
+        {
+            break;
+        }
+    }
+}
+
+void tiny_vicky_video_device::draw_mouse(uint32_t *row, bool enable_gamma, uint16_t line, uint16_t width, uint16_t height)
+{
+    uint8_t mouse_reg = m_iopage0_ptr[0xD6E0 - 0xC000];
```

Comment:
Is this “I/O page” actually shared RAM, or is it a bunch of memory-mapped registers?  If it’s memory-mapped registers, you should put proper read/write handlers in the address map rather than pretending it’s memory.